java/arcs/core/storage/DirectStoreMuxer.kt

```diff
@@ -83,37 +116,57 @@ class DirectStoreMuxer<Data : CrdtData, Op : CrdtOperation, T>(
     }
 
     /**
-     * Sends the provided [ProxyMessage] to the store backing the provided [referenceId].
+     * Sends the provided [ProxyMessage] to the store backing the provided [muxId].
      *
-     * A new store will be created for the [referenceId], if necessary.
+     * A new store will be created for the [muxId], if necessary.
      */
     suspend fun onProxyMessage(
-        message: ProxyMessage<Data, Op, T>,
-        referenceId: String
+        message: ProxyMessage<Data, Op, T>
     ): Boolean {
-        val (id, store) = store(referenceId)
+        val muxId = message.muxId!!
+        val (idSet, store) = store(muxId)
+
+        if (!idSet.contains(message.id)) registerToStore(message.id!!, muxId, idSet, store)
+
         val deMuxedMessage: ProxyMessage<Data, Op, T> = when (message) {
-            is SyncRequest -> SyncRequest(id)
-            is ModelUpdate -> ModelUpdate(message.model, id)
+            is SyncRequest -> SyncRequest(message.id)
+            is ModelUpdate -> ModelUpdate(message.model, message.id)
             is Operations -> if (message.operations.isNotEmpty()) {
-                Operations(message.operations, id)
+                Operations(message.operations, message.id)
             } else return true
         }
         return store.onProxyMessage(deMuxedMessage)
     }
 
-    /* internal */ suspend fun setupStore(referenceId: String): StoreRecord<Data, Op, T> {
+    /* internal */ suspend fun setupStore(muxId: String): StoreRecord<Data, Op, T> {
         val store = DirectStore.create<Data, Op, T>(
             StoreOptions(
-                storageKey = storageKey.childKeyWithComponent(referenceId),
+                storageKey = storageKey.childKeyWithComponent(muxId),
                 type = backingType
             )
         )
 
-        val id = store.on(callbackFactory(referenceId))
-
         // Return a new Record and add it to our local stores, keyed by muxId.
-        return StoreRecord(id, store)
+        return StoreRecord(mutableSetOf<Int>(), store)
+    }
+
+    @VisibleForTesting
+    suspend fun registerToStore(
+        id: Int,
+        muxId: String,
+        idSet: MutableSet<Int>,
+        store: DirectStore<Data, Op, T>
```

Comment:
I ran into a bug trying to do this. I believe calling `store(..)` inside `registerToStore(..)` causes a deadlock because both functions require a mutex lock.

Reply:
can you store() before acquiring the lock then?
suspend fun registerToStore(newId: Int, muxId: String) {
    val (idSet, store) = store(muxId)
     if (!idSet.contains(message.id)) {
         storeMutex.withLock {
              ....

Reply:
Ah yes, that works and looks cleaner. Thanks!